Systematic testing for the runtime

dipinhora opened the pull request:
The overall idea and some details of the implementation for systematic
testing has been shamelessly stolen from the `Verona` runtime (see:
https://github.com/microsoft/verona/blob/master/docs/explore.md#systematic-testing
for details). This implementation doesn't include replayable runtime
unit tests like `Verona`, but it sets a foundation for allowing
replayable runs of programs (and probably tests) for debugging
runtime issues such as backpressure/etc.

ponylang-main commented:
<!-- release-note-reminder-bot -->
Hi @dipinhora,

The changelog - added label was added to this pull request; all PRs with a changelog label need to have release notes included as part of the PR. If you haven't added release notes already, please do.

Release notes are added by creating a uniquely named file in the `.release-notes` directory. We suggest you call the file `4140.md` to match the number of this pull request.

The basic format of the release notes (using markdown) should be:

```
## Title

End user description of changes, why it's important,
problems it solves etc.

If a breaking change, make sure to include 1 or more
examples what code would look like prior to this change
and how to update it to work after this change.
```

Thanks.

jemc commented:
Discussed briefly during sync, but I haven't had the chance to review in detail yet.

It's exciting stuff.

SeanTAllen commented:
@dipinhora please let us know when you think this is ready for review.

dipinhora commented:
@SeanTAllen sorry, this in a good place in terms of setting the foundation for systematic testing and ready for review. i currently don't have any other changes pending although there can/should probably be follow up PRs to add calls to `SYSTEMATIC_TESTING_YIELD` in many more places and other enhancements (some of which are in comments as `TODO:`)..

SeanTAllen commented:
So at the point this, what level of usability is this at? What could we do with it? I feel like we need some documentation around that somewhere.

SeanTAllen commented on src/libponyrt/asio/asio.c:
```diff
@@ -42,6 +58,24 @@ void ponyint_asio_init(uint32_t cpu)
+    // create wait event objects
+    running_base.sleep_object = CreateEvent(NULL, FALSE, FALSE, NULL);
+#elif defined(USE_SCHEDULER_SCALING_PTHREADS)
+    // TODO: memtrack accounting
```
I'd like to see any TODO's be able to be easily found like...

TODO systematic testing:

Or some unique value and then opening an issue with a checklist of things to add/do that notes what those things are and notes the special todo value to look for.

dipinhora replied:
Replaced `TODO:` with `TODO systematic testing:`. Will open an issue with the checklist/string to search for once this PR is merged.

jemc commented on src/libponyrt/sched/systematic_testing.c:
```diff
@@ -0,0 +1,217 @@
+  srand((int)random_seed);
+
+  // initialize thead tracking array (should be max_threads + 1 to account for asio)
+  // TODO: memtrack tracking for these allocations
```
Can you include this enhancement in the current PR?

dipinhora replied:
done

jemc commented on src/libponyrt/asio/asio.c:
```diff
@@ -42,6 +58,24 @@ void ponyint_asio_init(uint32_t cpu)
+    // create wait event objects
+    running_base.sleep_object = CreateEvent(NULL, FALSE, FALSE, NULL);
+#elif defined(USE_SCHEDULER_SCALING_PTHREADS)
+    // TODO: memtrack accounting
```
Can you include this enhancement in the current PR?

dipinhora commented on src/libponyrt/sched/systematic_testing.h:
```diff
@@ -0,0 +1,61 @@
+
+#if defined(USE_SYSTEMATIC_TESTING)
+#if !defined(PLATFORM_IS_WINDOWS) && !defined(USE_SCHEDULER_SCALING_PTHREADS)
+pony_static_assert(false, "Systematic testing requires pthreads (USE_SCHEDULER_SCALING_PTHREADS) to be enabled!");
```
> Why?

Because it was simpler to not worry about `signals` and rely only on `pthreads` to start with. In theory, it should be possible to extend this to using `signals` instead of `pthreads` when `USE_SCHEDULER_SCALING_PTHREADS` is not defined.

> Shouldn't it be possible to write this in such a way that it works both with and without scaling enabled?

i don't understand this question. The current implementation already works whether or not `scaling` is disabled or not via the (`--ponynoscale` cli option). Can you clarify what you meant by this question?

SeanTAllen replied:
At the time Joe asked the question, he had forgotten that you can turn on scaling and then one of two options for how the scaling is done. When we reviewed during sync today I reminded him of that.

dipinhora replied:
Got it. Thanks for the info.

dipinhora commented:
> So at the point this, what level of usability is this at?

It functions and allows one to reproduce a deterministic run of a program compiled with the runtime.

> What could we do with it?

As an example, if someone has a test that has an intermittent failure (that is somehow related to timing of how actors are scheduled and run) they could recompile the test with systematic testing enabled and then run the test until it fails and then continually reproduce the failure by re-using the same `seed` via the `--ponysystematictestingseed <SEED_THAT_CAUSED_FAILURE>` cli argument. Then once the intermittent failure can be reliably reproduced, it should make it significantly easier to track down the root cause and fix the bug.

> I feel like we need some documentation around that somewhere.

More than happy to add it assuming you have some suggestions as to what and where.

dipinhora commented:
It could be useful to users of `ponyc` (like in the example scenario in my last comment) but i would expect it to get more use from developers of Pony as they enhance the runtime (i.e. changes to backpressure, changes to the message queue, changes to the objectmap, changes to GC, changes to the cycle detector, etc).

SeanTAllen commented:
@dipinhora i'm not sure where docs should go or what they should be. im open to ideas but i think without docs that people can refer to, it will be underused. Perhaps the contributors repo? Perhaps a new top-level document in this repo that gets referenced from BUILD.md. In this repo feels best as it can easily change as the functionality changes.

dipinhora commented:
@SeanTAllen i am all for adding in documentation for folks to make it easier to use the feature and i will think on what that could look like. However, i would prefer to get the changes merged even without comprehensive documentation to avoid a bitrot type scenario with the changes in this PR. Do you have any suggestions for what would be considered an adequate amount of documentation before this PR can be merged (assuming all other review concerns are also addressed)?

SeanTAllen commented:
I think information in BUILD on how to build and a new doc that details the basic of usage.

dipinhora commented:
i've updated `BUILD.md` and added a new `SYSTEMATIC_TESTING.md`. The new document isn't comprehensive but it sets a foundation that can be enhanced and built upon similar to how the rest of this PR sets a foundation for systematic testing that can be enhanced and built upon.

SeanTAllen commented:
@jemc based on Dipin's reply, is there anything you want to see done with this before we merge this initial version?

SeanTAllen commented:
@dipinhora seems like the final commit message should be updated? you have a lot of nice content in the release notes etc that suggest we could have a more meaningful commit message. If you agree, please update the first comment and I'll use that. If no, let me know. 

Either way, ping me and I'll squash and merge this whichever way you decide to go with the final commit message.

dipinhora commented:
@SeanTAllen first comment updated
